### PR TITLE
[jit] Add removeAttribute to `ClassType`

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28985 [quant][graphmode] Remove observer module after insert_quant_dequant
* #29048 [jit] Add unsafeRemoveAttr and unsafeRemoveSlot to ivalue::Object
* **#28984 [jit] Add removeAttribute to `ClassType`**

Summary:
Support removing an attribute in `ClassType`, `ClassType` is
considered as a low level API and user of this function
need to guarantee the safety of calling this method.

Test Plan:
build/bin/test_jit

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D18253776](https://our.internmc.facebook.com/intern/diff/D18253776)

### CODE REVIEW

File: torch/csrc/jit/script/class_type.cpp [hunk 1]
```diff
@@ -92,6 +92,22 @@ size_t ClassType::addAttribute(
   return slot;
 }
 
+void ClassType::removeAttribute(const std::string& name) {
+  std::vector<std::string> attributeNames;
+  std::vector<TypePtr> attributeTypes;
+  auto& names = attributeNames_;
+  auto& types = attributeTypes_;
+  TORCH_CHECK(std::find(names.begin(), names.end(), name) != names.end(), "Can't remove a non-existent attribute");
```
Comment (zdevito) [inline torch/csrc/jit/script/class_type.cpp]: Please use getAttributeSlot, it does the assertion and the find.

File: aten/src/ATen/core/jit_type.h [hunk 2]
```diff
@@ -1476,6 +1476,11 @@ struct CAFFE2_API ClassType : public NamedType {
       TypePtr type,
       bool is_parameter = false);
 
+  // [Internal Only] Remove attribute from the ClassType
+  // caller is responsible to make sure the attribute is
+  // no longer used
+  void removeAttribute(const std::string& name);
```
Comment (zdevito) [inline aten/src/ATen/core/jit_type.h]: Let's put the warning in the function name: `unsafeRemoveAttribute`. Let's also add what it means for the attribute to be no longer used: it is unsafe  to having existing allocations of this object around anymore, and any code that works on the attribute is now invalid. Only newly created code is valid again.

File: torch/csrc/jit/script/class_type.cpp [hunk 3]
```diff
@@ -92,6 +92,22 @@ size_t ClassType::addAttribute(
...
+  for (size_t i = 0; i < names.size(); ++i) {
```
Comment (zdevito) [inline torch/csrc/jit/script/class_type.cpp]: These can just be calls to attributeNames_.erase, etc.

**Review by zdevito — COMMENTED** [review]
See inline comments, also needs tests.

**Comment by jerryzh168** [comment]
@zdevito for test, I didn't find existing tests for `ClassType`, should I create a new test file?

**Comment by suo** [comment]
> @zdevito for test, I didn't find existing tests for `ClassType`, should I create a new test file?

There is test_class_import and test_class_parser, you can make a new file that looks like those

**Comment by jerryzh168** [comment]
please take a look again @zdevito @suo

**Review by zdevito — APPROVED** [review]
Looks good, thanks!

**Pull request closed without merging**